valence_anvil/src/region.rs

```diff
@@ -0,0 +1,394 @@
+            position,
+            header,
+        })
+    }
+
+    /// Get the last time the chunk was modified in seconds since epoch.
+    pub fn chunk_timestamp(&self, chunk_pos: ChunkPos) -> Option<ChunkTimestamp> {
+        self.header
+            .timestamp((chunk_pos.x & 31) as usize, (chunk_pos.z & 31) as usize)
+            .into_option()
+    }
+
+    async fn read_chunk_bytes(&self, chunk_pos: ChunkPos) -> Result<Option<Vec<u8>>, Error> {
+        let seek_pos = self
+            .header
+            .offset((chunk_pos.x & 31) as usize, (chunk_pos.z & 31) as usize);
+
+        let mut lock = self.source.lock().await;
+
+        lock.seek(SeekFrom::Start(seek_pos.offset() + self.offset))
+            .await?;
+
+        if seek_pos.len() == 0 {
+            return Ok(None);
+        }
+
+        let compressed_chunk_size = {
+            let mut buf = [0u8; 4];
+            lock.read_exact(&mut buf).await?;
+            BigEndian::read_u32(&buf) as usize
+        };
+
+        if compressed_chunk_size == 0 {
+            return Err(Error::DataFormatError(DataFormatError::InvalidChunkSize(
+                compressed_chunk_size,
+            )));
+        }
+
+        let compression = CompressionScheme::from_raw(lock.read_u8().await?)?;
+        let uncompressed_buffer = compression
+            .read_to_vec(&mut *lock, compressed_chunk_size - 1)
+            .await?;
+        Ok(Some(uncompressed_buffer))
+    }
+
+    pub(crate) async fn parse_chunks<I: IntoIterator<Item = ChunkPos>>(
+        &self,
+        world: &AnvilWorld,
+        positions: I,
+    ) -> Result<impl Iterator<Item = (ChunkPos, Option<UnloadedChunk>)>, Error> {
+        let mut results = Vec::<(ChunkPos, Option<UnloadedChunk>)>::new();
+
+        for pos in positions.into_iter() {
+            assert!(
+                self.position.contains(pos),
+                "Chunk position {:?} was not found in region {:?}",
+                pos,
+                self.position
+            );
+
+            let chunk_data = self.read_chunk_bytes(pos).await?;
+            if let Some(chunk_data) = chunk_data {
+                let mut nbt = valence::nbt::from_binary_slice(&mut chunk_data.as_slice())?.0;
+                match Self::parse_chunk_nbt(&mut nbt, world) {
+                    Err(Error::DataFormatError(DataFormatError::InvalidChunkState(..))) => {
+                        results.push((pos, None));
+                    }
+                    Err(e) => return Err(e),
+                    Ok(parsed_chunk) => {
+                        results.push((pos, Some(parsed_chunk)));
+                    }
+                }
+            } else {
+                results.push((pos, None));
+            }
+        }
+
+        Ok(results.into_iter())
+    }
+
+    //TODO: This function is very large and should be separated into dedicated
+    // functions at some point.
+    fn parse_chunk_nbt(nbt: &mut Compound, world: &AnvilWorld) -> Result<UnloadedChunk, Error> {
```

Comment:
I have removed the `From` traits from valence_nbt for now just to be on the safe side. A good implementation might be to return an error type instead of `None` as shown in https://github.com/rust-lang/rfcs/pull/2484.